### PR TITLE
Automl data creator support for pytorch

### PR DESCRIPTION
**Pull request by TheaperDeng**
support code for ray tune searcher and autoest ut should be added later.
some code inherited from #3894

### CODE REVIEW

File: pyzoo/test/zoo/automl/model/test_base_pytorch_model.py [hunk 1]
```diff
@@ -23,17 +24,55 @@
...
-    train_x, train_y = get_linear_data(2, 5, 1000)
-    val_x, val_y = get_linear_data(2, 5, 400)
+    train_x, train_y = get_linear_data(2, 5, train_size)
+    val_x, val_y = get_linear_data(2, 5, valid_size)
     data = {'x': train_x, 'y': train_y, 'val_x': val_x, 'val_y': val_y}
     return data
 
 
+class CustomDataset(Dataset):
+    def __init__(self, mode="train", train_size=1000, valid_size=400):
```
Comment (yushan111) [inline pyzoo/test/zoo/automl/model/test_base_pytorch_model.py:39]: def __init__(self, size)

File: pyzoo/test/zoo/automl/model/test_base_pytorch_model.py [hunk 2]
```diff
@@ -23,17 +24,55 @@
...
     def get_linear_data(a, b, size):
         x = np.arange(0, 10, 10 / size, dtype=np.float32)
         y = a*x + b
         return x, y
-    train_x, train_y = get_linear_data(2, 5, 1000)
-    val_x, val_y = get_linear_data(2, 5, 400)
+    train_x, train_y = get_linear_data(2, 5, train_size)
+    val_x, val_y = get_linear_data(2, 5, valid_size)
     data = {'x': train_x, 'y': train_y, 'val_x': val_x, 'val_y': val_y}
     return data
```
Comment (yushan111) [inline pyzoo/test/zoo/automl/model/test_base_pytorch_model.py:35]: only keep get_linear_data(size, a=2, b=5)

File: pyzoo/test/zoo/orca/automl/autoestimator/test_autoestimator_pytorch.py [hunk 3]
```diff
@@ -114,6 +160,26 @@ def test_fit(self):
...
+    def test_fit_data_creator(self):
+        auto_est = AutoEstimator.from_torch(model_creator=model_creator,
+                                            optimizer=get_optimizer,
+                                            loss=nn.BCELoss(),
+                                            logs_dir="/tmp/zoo_automl_logs",
+                                            resources_per_trial={"cpu": 2},
+                                            name="test_fit")
+        search_space = create_linear_search_space()
+        search_space.update({"train_size": hp.qrandint(800, 1000, q=2),
+                             "valid_size": hp.qrandint(400, 500, q=2)})
```
Comment (yushan111) [inline pyzoo/test/zoo/orca/automl/autoestimator/test_autoestimator_pytorch.py:172]: Do we need to tune "train_size" and "valid_size"? If not, we could set the values in dataloader

File: pyzoo/zoo/automl/model/base_pytorch_model.py [hunk 4]
```diff
@@ -108,15 +116,30 @@ def update_config():
             self._check_config(**tmp_config)
             self.config.update(config)
 
+        # get train_loader and validation_loader
+        if hasattr(data, '__call__'):
```
Comment (yushan111) [inline pyzoo/zoo/automl/model/base_pytorch_model.py:120]: import types
if isinstance(data, types.FunctionType)

File: pyzoo/zoo/automl/model/base_pytorch_model.py [hunk 5]
```diff
@@ -108,15 +116,30 @@ def update_config():
...
+            train_loader = data(self.config)
+            validation_loader = validation_data(self.config)
+        else:
```
Comment (yushan111) [inline pyzoo/zoo/automl/model/base_pytorch_model.py:123]: check data is a tuple of ndarray

File: pyzoo/zoo/automl/model/base_pytorch_model.py [hunk 6]
```diff
@@ -108,15 +116,30 @@ def update_config():
...
+            x, y, validation_data = PytorchBaseModel.covert_input(x, y, validation_data)
+            val_x = self._reshape_input(validation_data[0])
+            val_y = self._reshape_input(validation_data[1])
+            batch_size = self.config["batch_size"]
+            train_loader = DataLoader(TensorDataset(x, y),
+                                      batch_size=int(batch_size),
+                                      shuffle=True)
+            validation_loader = DataLoader(TensorDataset(val_x, val_y),
+                                           batch_size=int(batch_size),
+                                           shuffle=True)
```
Comment (yushan111) [inline pyzoo/zoo/automl/model/base_pytorch_model.py:135]:
```python 
def np_to_creator(data):
      def data_creator(config):
            # reshape input / covert input
            return Dataloader()
      return data_creator

train_data_creator = np_to_creator(data)
val_data_creator = np_to_creator(validation_data)

File: pyzoo/test/zoo/automl/model/test_base_pytorch_model.py [hunk 7]
```diff
@@ -23,17 +24,55 @@
...
 
 
+class CustomDataset(Dataset):
```
[the remark this answers is not in the document]
Comment (yushan111) [reply pyzoo/test/zoo/automl/model/test_base_pytorch_model.py:38]: Or we can reuse the `get_linear_data`

File: pyzoo/test/zoo/orca/automl/autoestimator/test_autoestimator_pytorch.py (hunk 3 above)
Comment (TheaperDeng) [reply pyzoo/test/zoo/orca/automl/autoestimator/test_autoestimator_pytorch.py]: yes, "train_size" can be a tunable hp. while for this test, it is only used to test if the data creator config can be tuned by the search_space.

Comment (yushan111) [reply pyzoo/test/zoo/orca/automl/autoestimator/test_autoestimator_pytorch.py]: But for most use cases, users may not try tuning "train_size" for better accuracy.. shall we just use "batch_size" and "shuffle" to verify whether data creator search works?

**Comment by TheaperDeng** [comment]
pr validation combine pass:
http://10.239.176.111:18888/job/ZOO-PR-Validation-Other/5610/
http://10.239.176.111:18888/job/ZOO-PR-Validation/5581/
